packages/flame_yarn/lib/src/parse/parse.dart

```diff
@@ -0,0 +1,788 @@
+        typeError(
+          'variable $variableName of type ${variableType.name} cannot be '
+          'assigned a value of type ${expression.type.name}',
+        );
+      }
+    } else {
+      final variableType = expression.type;
+      final dynamic initialValue = variableType is String
+          ? ''
+          : variableType is num
+              ? 0
+              : variableType is bool
+                  ? false
+                  : null;
+      project.variables.setVariable(variableName, initialValue);
+      assert(project.variables.getVariableType(variableName) == variableType);
+    }
+    if (assignmentToken != Token.operatorAssign) {
+      if (!variableExists) {
+        nameError('variable $variableName was not defined');
+      }
+      throw UnimplementedError();
+    }
+    take(Token.endCommand);
+    return SetCommand(variableName, expression);
+  }
+
+  Command parseUserDefinedCommand() {
+    throw UnimplementedError('user-defined commands are not supported yet');
+  }
+
+  static const List<Token> assignmentTokens = [
+    Token.operatorAssign,
+    Token.operatorDivideAssign,
+    Token.operatorMinusAssign,
+    Token.operatorModuloAssign,
+    Token.operatorMultiplyAssign,
+    Token.operatorPlusAssign,
+  ];
+
+  //#endregion
+
+  //#region Expression parsing
+  //----------------------------------------------------------------------------
+  // EXPRESSIONS
+  //
+  // Expressions are the mathematical notation used within the commands and
+  // interpolated text in order to perform calculations at runtime.
+  //
+  // See https://github.com/YarnSpinnerTool/YarnSpinner/blob/main/Documentation/Yarn-Spec.md#expressions
+  //----------------------------------------------------------------------------
+
+  /// Parses an expression starting from the current [position], and up to the
+  /// point where we encounter a token that cannot be interpreted as part of an
+  /// expression. Note that the startExpression / endExpressions tokens are not
+  /// consumed (nor required) by this method. This makes it suitable to use
+  /// this method to both extract an interpolated text expression, and a command
+  /// expression.
+  ///
+  /// If an expression cannot be parsed at the current location at all, the
+  /// [constVoid] will be returned.
+  Expression parseExpression() {
+    // We're using the Operator-Precedence parsing algorithm here, see
+    // https://en.wikipedia.org/wiki/Operator-precedence_parser
+    final lhs = parsePrimary();
+    return _parseExpressionImpl(lhs, 0);
+  }
+
+  /// This is an implementation function for the Operator-Precedence parsing
+  /// algorithm. It consumes expressions of the form `LHS op RHS op...`, where
+  /// the precedence of operators must be greater or equal to [minPrecedence].
+  /// The initial [lhs] sub-expression is provided, and the parsing position
+  /// should be at the start of the next operator.
+  Expression _parseExpressionImpl(Expression lhs, int minPrecedence) {
```

Comment:
`lhs` and `rhs` should preferably have better names.

Reply:
Maybe left `left` and `right` would be enough?

Reply:
It's a kinda of a standard terminology though when it comes to expressions. In fact, these terms even have their own wikipedia page: https://en.wikipedia.org/wiki/Sides_of_an_equation :)

Reply:
For sure, but is it really used that much in parser terminology when it is actually written code?
I can't remember us using that terminology when coding parsers in any uni CS courses, but that was quite some time ago so I might be wrong.

Reply:
And `lhs` in the context here, wouldn't it make more sense to be named just `expression`?
This discussion is quite some bike-shedding though and as you've seen I've already approved the PR.

Reply:
They did use this terminology [here](https://en.wikipedia.org/wiki/Operator-precedence_parser#Pseudocode) where I borrowed the algorithm from, so it seemed natural to me to use the same.